### PR TITLE
Refactor common code from SNSPowderReduction to utils

### PR DESCRIPTION
**Pull request by ckendrick**
~~Do not merge before #30205~~

**Description of work.**

Pulls some absorption correction functions which calculate the absorption correction and create a donor workspace out of `SNSPowderReduction` and move them into a new file in the `mantid.utils` module so that they can be reused in other places.

This also adds a doc page for `mantid.utils` with a brief explanation and example codes. A testing directory is also made to run unit tests on code added to `mantid.utils`. 

**To test:**

Run the `SNSPowderRedux` system test to make sure the code refactor did not break the algorithm, and run the new unit tests with `ctest -R python.utils`.

*There is no associated issue.*

The version of this into `ornl-next` is #30194

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by clayton-tom — COMMENTED** [review]
Code looks good but I'm getting a memory increase on the system tests (running on windows), I've rerun cmake and built in release, is there anything else that needs to be done?

File: Testing/SystemTests/tests/framework/AlignAndFocusPowderFromFilesTest.py
Comment (ckendrick) [reply Testing/SystemTests/tests/framework/AlignAndFocusPowderFromFilesTest.py]: Sorry, these were not related to this PR. This branch was rebased on `ornl-next` before those fixes got merged into `master` later that day, so these fixes got brought in to this branch. I just rebased it again, so this problem should be fixed now.

File: Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py
Comment (ckendrick) [reply Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py]: This is a good idea. I'm not sure if there is an existing place for constants in Python, so I just defined it in the utils file. It can be moved if there is a better place to put it

**Comment by ckendrick** [comment]
Thanks for catching the memory increase! I think it was caused by duplicated `Load` calls to get metadata, so I added an option to pass a workspace through to the absorption correction functions instead of having to have a `Load` in both places. There was also an unnecessary `Load` if no absorption correction was specified. I ran this on linux and the memory footprints for all the `SNSPowderRedux` tests were reduced back to their expected values. Let me know if this didn't fix it for you on Windows.

File: Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py
Comment (clayton-tom) [reply Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py]: Looks good to me

File: Testing/SystemTests/tests/framework/AlignAndFocusPowderFromFilesTest.py
Comment (clayton-tom) [reply Testing/SystemTests/tests/framework/AlignAndFocusPowderFromFilesTest.py]: Understood, thanks!

**Review by clayton-tom — COMMENTED** [review]
Thanks for those changes, however I'm still getting a memory increase on windows, I'm wondering if this is an issue on my end however as the Jenkins build doesn't seem to be picking it up?

**Comment by marshallmcdonnell** [comment]
@clayton-tom sorry to bother you but I was just going to gently check on the status of windows memory issue?
Asking since waiting for this to get merged for follow-on work.
Unfortunately, don't have any windows machines besides the Jenkins build servers to test otherwise.

No rush and thanks for reviewing this!

**Comment by clayton-tom** [comment]
@marshallmcdonnell @ckendrick  apologies for not getting this sorted before my time off. I've double checked that the memory increase is still happening on my end, but I also gave the tests a spin on the most recent master build and also saw the memory increase there, so it seems to be a preexisting problem. With this in mind, I appreciate this small issue may not be top of the priority list to sort out, so I'm happy to approve the PR unless you'd like to get a fix in as part of it. Happy New Year!

**Comment by marshallmcdonnell** [comment]
@clayton-tom Thanks for checking again and also the comparison to `master`. Sorry again to bother you over the holiday break. I'll leave it to @ckendrick. Yet, just my opinion, I believe the memory issue probably should be moved to a separate issue. Good catch! Maybe @martyngigg may have some insights? 

Thanks again!

**Comment by ckendrick** [comment]
@clayton-tom Thanks for checking the memory increase again! I think since you are seeing it on `master` as well, then it would probably be best to open an separate issue to investigate the memory increase.

**Comment by martyngigg** [comment]
It sounds to me like the memory increase on Windows is a separate issue as this code mostly runs on Linux and the issue already exists on master.

File: Framework/PythonInterface/mantid/utils/AbsorptionCorrUtils.py
Comment (clayton-tom) [inline Framework/PythonInterface/mantid/utils/AbsorptionCorrUtils.py:3]: One last thing - is it worth updating the year as this is a new file?

Comment (martyngigg) [reply Framework/PythonInterface/mantid/utils/AbsorptionCorrUtils.py:3]: Could I be slightly annoying and ask we rename the file to `absorptioncorrutils.py` for a more Pythonic naming convention inline with the rest of the `mantid` module?

**Review by clayton-tom — APPROVED** [review]
Thanks for those last changes 👍

**Pull request merged**
